.buildkite/bin/prh.sh

```diff
@@ -109,29 +110,69 @@ function watch_for_hang {
             ps -p $pid || continue
-            
             # Found a hung process; now kill it
             echo "Process $pid is (still) valid / running"
             echo ""
             echo 'List of hung processes dependent on $pid'
             echo 'ps -xo "%p %P %y %x %c" --sort ppid | grep $pid'
             ps -xo "%p %P %y %x %c" --sort ppid | grep $pid
             echo ""
-            echo "$tag KILL $pid !"
+            em='!'; echo "$tag KILL $pid$em"
             echo "kill $pid"
             kill $pid
             echo "$tag DONE"
+
+            # But that's not all!
+            # QDIR=/cad/cadence/EXT-ISR1_19.11.000_lnx86/tools.lnx86/extraction/bin/64bit
+            # ITMP=/sim/tmp/innovus/sim/tmp/innovus
+            # CMD1=r7arm-aha_buildkite-agent_MR0fEJ/tmp_qrc_QWq3Uu/qrc.cmd
+            # CMD2=r7arm-aha_buildkite-agent_MR0fEJ/tmp_qrc_QWq3Uu/__qrc.qrc.cmd
+            # GZ=$ITMP_temp_2934_r7arm-aha_buildkite-agent_MR0fEJ/tmp_qrc_QWq3Uu/qrc.def.gz
+            # 24253 pts/17   Sl+    0:13 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD1 $GZ
+            # 24461 pts/17   Sl+    0:15 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD2 $GZ
+            # 24823 pts/17   Sl+    0:06 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD2 $GZ
+            # 25109 pts/17   Sl+    0:06 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD2 $GZ
+            # 25594 pts/17   Sl+    0:06 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD2 $GZ
+            # 25595 pts/17   Sl+    0:06 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD2 $GZ
+            # 25596 pts/17   Sl+    0:06 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD2 $GZ
+            # 25597 pts/17   Sl+    0:06 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD2 $GZ
+            # 25599 pts/17   Sl+    0:06 $QDIR/qrc -cmd $ITMP_temp_2934_$CMD2 $GZ
+            echo "Haha no not even"
+            echo "kill dangling qrc -cmd jobs"; echo ""
+            ps ax | egrep temp_${pid}_ | grep -v grep
+            bad_pids=`ps ax | egrep temp_${pid}_ | grep -v grep | awk '{print $1}'`
+            echo kill $bad_pids
+            kill $bad_pids; echo ""
+            
+            # ...and still more processes to kill :(
+            # QDIR=/cad/cadence/EXT-ISR1_19.11.000_lnx86/tools.lnx86/extraction/bin/64bit
+            # 24360 pts/17   S+     1:20 $QDIR/qrc -srv r7arm-aha 36989 0 -1
+            # 24462 pts/17   S+     1:18 $QDIR/qrc -srv r7arm-aha 36989 1 -1
+            # 24671 pts/17   S+     1:15 $QDIR/qrc -srv r7arm-aha 36989 2 -1
+            # 24824 pts/17   S+     1:16 $QDIR/qrc -srv r7arm-aha 36989 3 -1
+            # 25012 pts/17   S+     1:15 $QDIR/qrc -srv r7arm-aha 36989 4 -1
+            # 25110 pts/17   S+     1:16 $QDIR/qrc -srv r7arm-aha 36989 5 -1
+            # 25286 pts/17   S+     1:15 $QDIR/qrc -srv r7arm-aha 36989 6 -1
+            # 25478 pts/17   S+     1:16 $QDIR/qrc -srv r7arm-aha 36989 7 -1
+            echo "kill dangling qrc -srv jobs"
+            ps x | grep 'qrc..srv'; echo ""
+            bad_pids=`ps x | grep 'qrc..srv' | awk '{print $1}'`
```

Comment:
Does this need to be made specific to the particular pid like you did for the qrc -cmd jobs?

Reply:
Good catch! In a perfect world, this would be the case. Here's why I didn't do that:

The '-cmd' files have the pid embedded in the filename e.g. if a job was spawned by process 2934 the filename is `..._temp_2934_...` and so I can extract the jobs specific to that pid. The '-srv' command does not appear to have pid embedded, so I just, you know, kill 'em all and let God sort it out. The downside is that the buildkite agent may try to kill jobs that don't belong to it. The upside is that linux won't let it do that, because of permissions, so the damage is limited....